zetaclient/evm/inbounds.go

```diff
@@ -59,221 +59,157 @@ func (ob *ChainClient) ObserveTrackerSuggestions() error {
 		return err
 	}
 	for _, tracker := range trackers {
-		ob.logger.ExternalChainWatcher.Info().Msgf("checking tracker with hash :%s and coin-type :%s ", tracker.TxHash, tracker.CoinType)
+		// query tx and receipt
+		tx, _, err := ob.TransactionByHash(tracker.TxHash)
+		if err != nil {
+			return errors.Wrapf(err, "error getting transaction for intx %s chain %d", tracker.TxHash, ob.chain.ChainId)
+		}
+		receipt, err := ob.evmClient.TransactionReceipt(context.Background(), ethcommon.HexToHash(tracker.TxHash))
+		if err != nil {
+			return errors.Wrapf(err, "error getting receipt for intx %s chain %d", tracker.TxHash, ob.chain.ChainId)
+		}
+		ob.logger.ExternalChainWatcher.Info().Msgf("checking tracker for intx %s chain %d", tracker.TxHash, ob.chain.ChainId)
+
+		// check and vote on inbound tx
 		switch tracker.CoinType {
 		case common.CoinType_Zeta:
-			ballotIdentifier, err := ob.CheckReceiptForCoinTypeZeta(tracker.TxHash, true)
-			if err != nil {
-				return err
-			}
-			ob.logger.ExternalChainWatcher.Info().Msgf("Vote submitted for inbound Tracker,Chain : %s,Ballot Identifier : %s, coin-type %s", ob.chain.ChainName, ballotIdentifier, common.CoinType_Zeta.String())
+			_, err = ob.CheckNVoteInboundTokenZeta(tx, receipt, true)
 		case common.CoinType_ERC20:
-			ballotIdentifier, err := ob.CheckReceiptForCoinTypeERC20(tracker.TxHash, true)
-			if err != nil {
-				return err
-			}
-			ob.logger.ExternalChainWatcher.Info().Msgf("Vote submitted for inbound Tracker,Chain : %s,Ballot Identifier : %s, coin-type %s", ob.chain.ChainName, ballotIdentifier, common.CoinType_ERC20.String())
+			_, err = ob.CheckNVoteInboundTokenERC20(tx, receipt, true)
 		case common.CoinType_Gas:
-			ballotIdentifier, err := ob.CheckReceiptForCoinTypeGas(tracker.TxHash, true)
-			if err != nil {
-				return err
-			}
-			ob.logger.ExternalChainWatcher.Info().Msgf("Vote submitted for inbound Tracker,Chain : %s,Ballot Identifier : %s, coin-type %s", ob.chain.ChainName, ballotIdentifier, common.CoinType_Gas.String())
+			_, err = ob.CheckNVoteInboundTokenGas(tx, receipt, true)
+		default:
+			return fmt.Errorf("unknown coin type %s for intx %s chain %d", tracker.CoinType, tx.Hash, ob.chain.ChainId)
+		}
+		if err != nil {
+			return errors.Wrapf(err, "error checking and voting for intx %s chain %d", tx.Hash, ob.chain.ChainId)
 		}
 	}
 	return nil
 }
 
-func (ob *ChainClient) CheckReceiptForCoinTypeZeta(txHash string, vote bool) (string, error) {
-	addrConnector, connector, err := ob.GetConnectorContract()
-	if err != nil {
-		return "", err
+// CheckNVoteInboundTokenZeta checks and votes on the given inbound Zeta token
+func (ob *ChainClient) CheckNVoteInboundTokenZeta(tx *ethrpc.Transaction, receipt *ethtypes.Receipt, vote bool) (string, error) {
```

Comment:
Is the vote flag only used for testing? You can probably use the zetabridge stub I added in my PR #1809

Reply:
The `false` flag is also used by zetaclientd's `debug` command [these 3 lines](https://github.com/zeta-chain/node/blob/feat-observe-blob-transaction/cmd/zetaclientd/debug.go#L158)